clients/src/test/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManagerTest.java

```diff
@@ -455,6 +468,192 @@ public void testRequestFails_AuthenticationException() {
+        when(subscriptionState.partitionsNeedingValidation(time.milliseconds())).thenReturn(Collections.singleton(TEST_PARTITION_1));
+        when(subscriptionState.position(any())).thenReturn(position, position);
+        NodeApiVersions nodeApiVersions = NodeApiVersions.create();
+        when(apiVersions.get(leader.idString())).thenReturn(nodeApiVersions);
+    }
+
+    private void testResetPositionsSuccessWithLeaderEpoch(Metadata.LeaderAndEpoch leaderAndEpoch) {
+        TopicPartition tp = TEST_PARTITION_1;
+        Node leader = LEADER_1;
+        OffsetResetStrategy strategy = OffsetResetStrategy.EARLIEST;
+        long offset = 5L;
+        Map<TopicPartition, OffsetAndTimestamp> expectedOffsets = Collections.singletonMap(tp,
+                new OffsetAndTimestamp(offset, 1L, leaderAndEpoch.epoch));
+        when(subscriptionState.partitionsNeedingReset(time.milliseconds())).thenReturn(Collections.singleton(tp));
+        when(subscriptionState.resetStrategy(any())).thenReturn(strategy);
+        expectSuccessfulRequest(Collections.singletonMap(tp, leader));
+
+        requestManager.resetPositionsIfNeeded();
+        assertEquals(1, requestManager.requestsToSend());
+
+        // Reset positions response with offsets
+        when(metadata.currentLeader(tp)).thenReturn(testLeaderEpoch(leader, leaderAndEpoch.epoch));
+        NetworkClientDelegate.PollResult pollResult = requestManager.poll(time.milliseconds());
+        NetworkClientDelegate.UnsentRequest unsentRequest = pollResult.unsentRequests.get(0);
+        ClientResponse clientResponse = buildClientResponse(unsentRequest, expectedOffsets);
+        clientResponse.onComplete();
+        assertTrue(unsentRequest.future().isDone());
+        assertFalse(unsentRequest.future().isCompletedExceptionally());
+    }
+
+    @Test
+    public void testResetPositionsThrowsPreviousException() {
+        when(subscriptionState.partitionsNeedingReset(time.milliseconds())).thenReturn(Collections.singleton(TEST_PARTITION_1));
+        when(subscriptionState.resetStrategy(any())).thenReturn(OffsetResetStrategy.EARLIEST);
+        expectSuccessfulRequest(Collections.singletonMap(TEST_PARTITION_1, LEADER_1));
+
+        requestManager.resetPositionsIfNeeded();
+
+        // Reset positions response with TopicAuthorizationException
+        NetworkClientDelegate.PollResult res = requestManager.poll(time.milliseconds());
+        NetworkClientDelegate.UnsentRequest unsentRequest = res.unsentRequests.get(0);
+        ClientResponse clientResponse = buildClientResponseWithErrors(
+                unsentRequest, Collections.singletonMap(TEST_PARTITION_1, Errors.TOPIC_AUTHORIZATION_FAILED));
+        clientResponse.onComplete();
+
+        assertTrue(unsentRequest.future().isDone());
+        assertFalse(unsentRequest.future().isCompletedExceptionally());
+
+        verify(subscriptionState).requestFailed(any(), anyLong());
+        verify(metadata).requestUpdate(false);
```

Comment:
TOPIC_AUTHORIZATION_FAILED is not a retriable error. So, intuitively, we shouldn't need to refresh the metadata because of that. Does the old consumer code do that too?

Reply:
You're right this one is not retriable, but the metadata update request is done for all failures in the reset, regardless of the distinction of retriable/non-retriable (it was the logic in the old consumer too, [here](https://github.com/apache/kafka/blob/4be174f5c1eed700411f0806a2eb391cee3f1177/clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetFetcherUtils.java#L311))